M2/Macaulay2/packages/SparseResultants.m2

```diff
@@ -462,6 +463,210 @@ genericMultihomogeneousPolynomial (VisibleList,VisibleList) := o -> (k,d) -> (
+    K := M#"ring";
+    X := M#"varsMultilinearForm"; 
+    n' := for i to #n-1 list n_(l_i);
+    X' := for i to #n-1 list X_(l_i);
+    R := K[flatten X'];
+    X' = apply(X',X0->apply(X0,u -> sub(u,R)));
+    multidimensionalMatrix makeList(sub(M#"multilinearForm",R),X')
+);
+
+makeRing = method();
+makeRing (List) := (L) -> (
+    if #L == 0 then error "expected a nonempty list";
+    n := {#L};
+    L' := first L;
+    while instance(L',List) do (
+        if #L' == 0 then error "expected nonempty lists";
+        n = prepend(#L',n);
+        L' = first L';
+    );
+    for i to #n-2 do L = flatten L;
+    K := ring matrix{L};
+    gensRing0(K,n)
+);
+
+makeMultilinearForm = method();
+makeMultilinearForm (List,List) := (L,X) -> (
+    if not instance(first L,List) then return ((matrix{last X})*transpose(matrix{L}))_(0,0);
+    X' := X_{0 .. #X-2};
+    return makeMultilinearForm(apply(L,l -> makeMultilinearForm(l,X')),X);
+);
+
+makeList = method();
+makeList (RingElement,List) := (F,X) -> (
+    K := coefficientRing ring F;
+    R := K[X_0]; for i from 1 to #X-1 do R = R[X_i];
+    C := sub(F,R);
+    for i to #X-1 do C = coeffs C; 
+    return C;
+);
+
+coeffs = method();
+coeffs (RingElement) := (F) -> flatten entries sub(last coefficients(matrix F,Monomials=>vars ring F),coefficientRing ring F);
+coeffs (List) := (L) -> apply(L,F -> coeffs F);
+
+applyToEntries = method();
+applyToEntries (List,Function) := (M,f) -> (
+    ff := method();
+    ff Thing := t -> f(t);
+    ff List := l -> apply(l,ff);
+    ff(M) 
+);
+applyToEntries (MultidimensionalMatrix,Function) := (M,f) -> multidimensionalMatrix applyToEntries(M#"list",f);
+
+genericMultidimensionalMatrix = method(TypicalValue => MultidimensionalMatrix, Options => {CoefficientRing => ZZ});
+genericMultidimensionalMatrix (VisibleList,Boolean) := o -> (n,b) -> (
+    F := genericMultihomogeneousPolynomial(toSequence n,#n:1,CoefficientRing=>o.CoefficientRing);
+    K := coefficientRing coefficientRing ring F;
+    if b then F = sub(sub(F,apply(gens coefficientRing ring F,u -> u => random(K))),K[gens ring F]);
+    x := gens ring F;
+    X := {};
+    for i to #n-1 do (
+        X = append(X,x_{0 .. n_i -1});
+        x = x_{n_i .. #x-1}
+    );
+    multidimensionalMatrix makeList(F,X)
+);
+genericMultidimensionalMatrix (VisibleList) := o -> (n) -> genericMultidimensionalMatrix(n,false,CoefficientRing=>o.CoefficientRing);
+
+randomMultidimensionalMatrix = method(TypicalValue => MultidimensionalMatrix, Options => {CoefficientRing => ZZ});
+randomMultidimensionalMatrix (VisibleList) := o -> (n) -> genericMultidimensionalMatrix(n,true,CoefficientRing=>o.CoefficientRing);
+
+GENSRING0 := local GENSRING0;
+
+gensRing0 = method();
+gensRing0 (Ring,List) := (K,n) -> (
+    if instance(GENSRING0_(K,n),List) then return GENSRING0_(K,n);
+    x := apply(#n,i -> getSymbol("x"|toString(i)));
```

Comment:
All this stuff with `GENSRING` can probably be replaced by a single use of `memoize`.

Reply:
Yes, changed. I wasn't aware of this useful method.